Model print current values

ialxn opened the pull request:
Fixes #1543. Actually, it is probably also an enhancement

ialxn committed:
Enhancement: also print current values of parameter with attribute __iter__

Simplify (reverse) the if/else logic.

Signed-off-by: Ivo Alxneit <ivo.alxneit@psi.ch>

francisco-dlp commented:
I would make it optional by e.g. adding a ``print_multivalue`` (I couldn't come up with a better name, maybe you could?) keyword argument to the function as there are some parameters with very "long" values.

ialxn committed:
Make printing of parameters with attribute __iter__ controlled by user.

These parameters potentially have many members. So make it possible to
skip the output by setting the keyword `skip_multi` to True.

Signed-off-by: Ivo Alxneit <ivo.alxneit@psi.ch>

francisco-dlp commented on hyperspy/model.py:
```diff
@@ -1480,7 +1483,11 @@ def print_current_values(self, only_free=True):
+                        if skip_multi:
+                            for idx in range(len(parameter.value)):
+                                print("\t\ta[%d]\t%g" % (idx, parameter.value[idx]))
+                    else:
```
This else block should be indented furhter to mach the previous ``if``.

ialxn replied:
No, this should be correct. The `else:` block concerns parameters that do not have the attribute `__iter__` set. The `if skip_multi:` is only applied to iterable parameters.

francisco-dlp replied:
You are right! Maybe the issue is that it should be ``if not skip_multi``?

francisco-dlp replied:
Also, wouldn't it be better to print the parameter name instead of ``a``? This is what I get when using ``skip_multi=True`` (should be ``False``) on an EELS model:

```python


Components	Parameter	Value
PowerLaw
		A	40
		r	-0
Fe_M3
		a[0]	0
		a[1]	0
		a[2]	0
		a[3]	0
		a[4]	0
		a[5]	0
		a[6]	0
		a[7]	0
		a[8]	0
		a[9]	0
		a[10]	0
		a[11]	0
		a[12]	0
		a[13]	0
		a[14]	0
		a[15]	0
		a[16]	0
		a[17]	0
		a[18]	0
		a[19]	0
		a[20]	0
		a[21]	0
		a[22]	0
		a[23]	0
		a[24]	0
		intensity	1
```

ialxn replied:
I'll check to get the parameter name printed and the `skip_multi` issue. But I guess, the parameter values (all zeros) are correct.

ialxn replied:
Yes, of course, it should be `if not skip_multi`.

Otherwise I am a bit at a loss. I only realized now that the `__iter__`able parameters can be more complex, like dictionaries as in the EELS model. Is there a generic way to iterate through the parameters that catch tuples, arrays, ..., dictionaries? Do I have to deal with things like dictionaries of dictionaries? Or a list that contains a dictionary and a list?

francisco-dlp replied:
``value`` can only contain numbers or a tuple of numbers, so the way you did it should be fine.

francisco-dlp commented:
#1549
